go/chat/partsource.go

```diff
@@ -0,0 +1,108 @@
+package chat
+
+import (
+	"context"
+
+	"github.com/keybase/client/go/chat/globals"
+	"github.com/keybase/client/go/chat/types"
+	"github.com/keybase/client/go/chat/utils"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/chat1"
+	"github.com/keybase/client/go/protocol/gregor1"
+)
+
+type partDiskStorage struct {
+	Uids []gregor1.UID
+	Hash string
+}
+
+type CachingParticipantSource struct {
+	globals.Contextified
+	utils.DebugLabeler
+
+	locktab *libkb.LockTable
+	ri      func() chat1.RemoteInterface
+}
+
+var _ types.ParticipantSource = (*CachingParticipantSource)(nil)
+
+func NewCachingParticipantSource(g *globals.Context, ri func() chat1.RemoteInterface) *CachingParticipantSource {
+	return &CachingParticipantSource{
+		Contextified: globals.NewContextified(g),
+		DebugLabeler: utils.NewDebugLabeler(g.ExternalG(), "CachingParticipantSource", false),
+		locktab:      libkb.NewLockTable(),
+		ri:           ri,
+	}
+}
+
+func (s *CachingParticipantSource) Get(ctx context.Context, conv types.RemoteConversation) (res []gregor1.UID, err error) {
+	defer s.Trace(ctx, func() error { return err }, "Get")()
+	ch := s.GetNonblock(ctx, conv)
+	for r := range ch {
+		res = r.Uids
+	}
+	return res, nil
+}
+
+func (s *CachingParticipantSource) dbKey(convID chat1.ConversationID) libkb.DbKey {
+	return libkb.DbKey{
+		Typ: libkb.DBChatParticipants,
+		Key: convID.String(),
+	}
+}
+
+func (s *CachingParticipantSource) GetNonblock(ctx context.Context, conv types.RemoteConversation) (resCh chan types.ParticipantResult) {
+	resCh = make(chan types.ParticipantResult, 1)
+	go func(ctx context.Context) {
+		defer s.Trace(ctx, func() error { return nil }, "GetNonblock")()
+		defer close(resCh)
+		lock := s.locktab.AcquireOnName(ctx, s.G(), conv.ConvIDStr.String())
+		defer lock.Release(ctx)
+
+		switch conv.GetMembersType() {
+		case chat1.ConversationMembersType_TEAM:
+			// handle team separately in here
+		default:
+			resCh <- types.ParticipantResult{Uids: conv.Conv.Metadata.AllList}
+			return
+		}
+
+		// load local first and send to channel
+		var local partDiskStorage
+		found, err := s.G().GetKVStore().GetInto(&local, s.dbKey(conv.GetConvID()))
```

Comment:
better to store as msgpack instead of json `GetIntoMsgpack`

Reply:
also, this should be in encrypted db

Reply:
Ah true, I'll change to encrypted db